README
NAME
    MooseX::AttributeTags - tag your Moose attributes

SYNOPSIS
       package User;
   
       use Moose;
       use MooseX::Types::Moose 'Bool';
       use MooseX::AttributeTraits (
          SerializationStyle => [
             hidden => Bool,
          ],
       );
   
       has username => (
          traits => [ SerializationStyle ],
          is     => 'ro',
          hidden => 0,
       );
   
       has password => (
          traits => [ SerializationStyle ],
          is     => 'rw',
          hidden => 1,
       );

DESCRIPTION
    MooseX::AttributeTags is a factory for attribute traits. All the work is
    done in the import method.

  Methods
    `import(@optlist)`
        The option list is a list of trait names to create (which will be
        exported to the caller package as constants).

        Each trait name may be optionally followed by an arrayref of
        attributes to be created within the trait. (In the SYNOPSIS, the
        "SerializationStyle" trait gets an attribute called "hidden".)

        Each attribute may be optionally followed by *one* of:

        *   A coderef which provides a default value for the attribute.

        *   A type constraint object (such as those provided by
            Types::Standard or MooseX::Types; not a type constraint string) to
            validate the attribute.

        *   An arrayref or hashref providing options similar to those given to
            Moose's `has` keyword.

BUGS
    Please report any bugs to
    <http://rt.cpan.org/Dist/Display.html?Queue=MooseX-AttributeTags>.

SEE ALSO
    Moose.

AUTHOR
    Toby Inkster <tobyink@cpan.org>.

COPYRIGHT AND LICENCE
    This software is copyright (c) 2013 by Toby Inkster.

    This is free software; you can redistribute it and/or modify it under the
    same terms as the Perl 5 programming language system itself.

DISCLAIMER OF WARRANTIES
    THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
    WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
    MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.



Changes
MooseX-AttributeTags
====================

Created:      2013-08-12
Home page:    <https://metacpan.org/release/MooseX-AttributeTags>
Bug tracker:  <http://rt.cpan.org/Dist/Display.html?Queue=MooseX-AttributeTags>
Maintainer:   Toby Inkster (TOBYINK) <tobyink@cpan.org>

0.002	2014-10-03

 [ Packaging ]
 - Add missing dependency on Test::Fatal.

0.001	2013-08-12	Initial release


Makefile.PL
use strict;
use ExtUtils::MakeMaker 6.17;

my $EUMM = eval( $ExtUtils::MakeMaker::VERSION );

my $meta = {
  "abstract"       => "tag your Moose attributes",
  "author"         => ["Toby Inkster (TOBYINK) <tobyink\@cpan.org>"],
  "dynamic_config" => 0,
  "generated_by"   => "Dist::Inkt::Profile::TOBYINK version 0.023, CPAN::Meta::Converter version 2.140640",
  "keywords"       => [],
  "license"        => ["perl_5"],
  "meta-spec"      => {
                        url => "http://search.cpan.org/perldoc?CPAN::Meta::Spec",
                        version => 2,
                      },
  "name"           => "MooseX-AttributeTags",
  "no_index"       => { directory => ["eg", "examples", "inc", "t", "xt"] },
  "prereqs"        => {
                        configure => { requires => { "ExtUtils::MakeMaker" => 6.17 } },
                        develop   => { recommends => { "Dist::Inkt" => 0.001 } },
                        runtime   => {
                                       requires => {
                                         "Data::OptList" => 0,
                                         "Moose" => "2.0600",
                                         "MooseX::Role::Parameterized" => 0,
                                         "perl" => "5.008000",
                                       },
                                     },
                        test      => { requires => { "Test::Fatal" => 0.007, "Test::More" => 0.96 } },
                      },
  "provides"       => {
                        "MooseX::AttributeTags"        => { file => "lib/MooseX/AttributeTags.pm", version => 0.002 },
                        "MooseX::AttributeTags::PRole" => { file => "lib/MooseX/AttributeTags/PRole.pm", version => 0.002 },
                      },
  "release_status" => "stable",
  "resources"      => {
                        bugtracker   => {
                                          web => "http://rt.cpan.org/Dist/Display.html?Queue=MooseX-AttributeTags",
                                        },
                        homepage     => "https://metacpan.org/release/MooseX-AttributeTags",
                        license      => ["http://dev.perl.org/licenses/"],
                        repository   => {
                                          type => "git",
                                          url  => "git://github.com/tobyink/p5-moosex-attributetags.git",
                                          web  => "https://github.com/tobyink/p5-moosex-attributetags",
                                        },
                        x_identifier => "http://purl.org/NET/cpan-uri/dist/MooseX-AttributeTags/project",
                      },
  "version"        => 0.002,
};

my %dynamic_config;

my %WriteMakefileArgs = (
	ABSTRACT   => $meta->{abstract},
	AUTHOR     => ($EUMM >= 6.5702 ? $meta->{author} : $meta->{author}[0]),
	DISTNAME   => $meta->{name},
	VERSION    => $meta->{version},
	EXE_FILES  => [ map $_->{file}, values %{ $meta->{x_provides_scripts} || {} } ],
	NAME       => do { my $n = $meta->{name}; $n =~ s/-/::/g; $n },
	test       => { TESTS => "t/*.t" },
	%dynamic_config,
);

$WriteMakefileArgs{LICENSE} = $meta->{license}[0] if $EUMM >= 6.3001;

sub deps
{
	my %r;
	for my $stage (@_)
	{
		for my $dep (keys %{$meta->{prereqs}{$stage}{requires}})
		{
			next if $dep eq 'perl';
			my $ver = $meta->{prereqs}{$stage}{requires}{$dep};
			$r{$dep} = $ver if !exists($r{$dep}) || $ver >= $r{$dep};
		}
	}
	\%r;
}

my ($build_requires, $configure_requires, $runtime_requires, $test_requires);
if ($EUMM >= 6.6303)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{TEST_REQUIRES}      ||= deps('test');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');
}
elsif ($EUMM >= 6.5503)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build', 'test');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');	
}
elsif ($EUMM >= 6.52)
{
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime', 'build', 'test');	
}
else
{
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('configure', 'build', 'test', 'runtime');	
}

{
	my ($minperl) = reverse sort(
		grep defined && /^[0-9]+(\.[0-9]+)?$/,
		map $meta->{prereqs}{$_}{requires}{perl},
		qw( configure build runtime )
	);
	
	if (defined($minperl))
	{
		die "Installing $meta->{name} requires Perl >= $minperl"
			unless $] >= $minperl;
		
		$WriteMakefileArgs{MIN_PERL_VERSION} ||= $minperl
			if $EUMM >= 6.48;
	}
}

sub FixMakefile
{
	return unless -d 'inc';
	my $file = shift;
	
	local *MAKEFILE;
	open MAKEFILE, "< $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	my $makefile = do { local $/; <MAKEFILE> };
	close MAKEFILE or die $!;
	
	$makefile =~ s/\b(test_harness\(\$\(TEST_VERBOSE\), )/$1'inc', /;
	$makefile =~ s/( -I\$\(INST_ARCHLIB\))/ -Iinc$1/g;
	$makefile =~ s/( "-I\$\(INST_LIB\)")/ "-Iinc"$1/g;
	$makefile =~ s/^(FULLPERL = .*)/$1 "-Iinc"/m;
	$makefile =~ s/^(PERL = .*)/$1 "-Iinc"/m;
	
	open  MAKEFILE, "> $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	print MAKEFILE $makefile or die $!;
	close MAKEFILE or die $!;
}

my $mm = WriteMakefile(%WriteMakefileArgs);
FixMakefile($mm->{FIRST_MAKEFILE} || 'Makefile');
exit(0);

